Fix __array_function__ dispatching for tril/triu

pentschev opened the pull request:
This PR fixes `__array_function__` dispatching for `tril`/`triu`, which broke for downstream libraries such as CuPy after #6997 .

- [ ] Closes #7324 , closes #7452 
- [ ] Tests added / passed
- [ ] Passes `black dask` / `flake8 dask`

pentschev commented on dask/array/creation.py:
```diff
@@ -677,17 +680,17 @@ def _diag_len(dim1, dim2, offset):
 @derived_from(np)
-def tri(N, M=None, k=0, dtype=float, chunks="auto"):
+def tri(N, M=None, k=0, dtype=float, chunks="auto", like=None):
     _min_int = np.lib.twodim_base._min_int
 
     if M is None:
         M = N
 
     chunks = normalize_chunks(chunks, shape=(N, M), dtype=dtype)
 
-    m = greater_equal.outer(
-        arange(N, chunks=chunks[0][0], dtype=_min_int(0, N)),
-        arange(-k, M - k, chunks=chunks[1][0], dtype=_min_int(-k, M - k)),
+    m = greater_equal(
+        arange(N, chunks=chunks[0][0], dtype=_min_int(0, N), like=like).reshape(1, N).T,
+        arange(-k, M - k, chunks=chunks[1][0], dtype=_min_int(-k, M - k), like=like),
```
Because CuPy doesn't support `greater_equal.outer`, I rewrote this in terms of a matrix transpose. I don't know if this imposes any performance impacts for NumPy-based use cases, if that seems like a problem please speak up @Illviljan .

Illviljan replied:
I haven't done any timeits but I imagine this is similar to what's done in numpy in the end so I'm not particularly worried. It should be an improvement over the previous dask implementation anyway.

pentschev replied:
Thanks for confirming.

Illviljan reviewed (commented):
Hello, I added some comments. 

Although it might be necessary I think it's a little unfortunate that you had to reimagine the implementation a bit. When I added it was pretty much a straight copy/paste from numpy. It made it very easy to review and see through possible errors (although not the cupy errors apparently...). But the numpy implementation has already changed since I copy/pasted it so keeping up with upstream is not easy.

pentschev commented on dask/array/routines.py:
```diff
@@ -1721,18 +1721,30 @@ def average(a, axis=None, weights=None, returned=False):
 def tril(m, k=0):
-    m = asarray(m)
-    mask = tri(*m.shape[-2:], k=k, dtype=bool, chunks=m.chunks[-2:])
+    from .utils import asarray_safe
```
Done.

pentschev commented on dask/array/routines.py:
```diff
@@ -1721,18 +1721,30 @@ def average(a, axis=None, weights=None, returned=False):
 def triu(m, k=0):
-    m = asarray(m)
-    mask = tri(*m.shape[-2:], k=k - 1, dtype=bool, chunks=m.chunks[-2:])
+    from .utils import asarray_safe
```
Done.

pentschev commented on dask/array/routines.py:
```diff
@@ -1721,18 +1721,30 @@ def average(a, axis=None, weights=None, returned=False):
+        *m.shape[-2:], k=k, dtype=bool, chunks=m.chunks[-2:], like=meta_from_array(m)
+    )
+
+    return where(mask, m, zeros_like_safe(m, shape=(1,)))
```
Yes, all the `*_safe` functions are used to satisfy older NumPy versions which don't support `__array_function__`, thus in this case falling back to regular `np.zeros` if `np.zeros_like` fails.

pentschev commented:
> Hello, I added some comments.

Thank you! They should all have been addressed now.

> Although it might be necessary I think it's a little unfortunate that you had to reimagine the implementation a bit. When I added it was pretty much a straight copy/paste from numpy. It made it very easy to review and see through possible errors (although not the cupy errors apparently...). But the numpy implementation has already changed since I copy/pasted it so keeping up with upstream is not easy.

Yes, supporting multiple backends can come with this sort of adapting when some functionality isn't supported. With that said, I'm not really worried about Dask having a diverging implementation to that of NumPy, due to its different requirements and different project paces it's generally impractical to keep everything in sync. Regardless of that, the new code is indeed much simpler than it used to be, thanks for working on that!

jakirkham reviewed (approved):
Thanks Peter! 😄

pentschev commented:
Thanks John for merging!
